src/simulator/halo_simulator.py

```diff
@@ -0,0 +1,196 @@
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Simulates the Halo system."""
+
+from typing import List
+from typing import Set
+from typing import Tuple
+
+import numpy as np
+
+from wfa_cardinality_estimation_evaluation_framework.estimators.same_key_aggregator import (
+    StandardizedHistogramEstimator,
+)
+from wfa_cardinality_estimation_evaluation_framework.common.noisers import (
+    LaplaceMechanism,
+)
+from wfa_cardinality_estimation_evaluation_framework.estimators.vector_of_counts import (
+    VectorOfCounts,
+)
+
+from wfa_planning_evaluation_framework.data_generators.data_set import DataSet
+from wfa_planning_evaluation_framework.models.reach_curve import ReachCurve
+from wfa_planning_evaluation_framework.models.reach_point import ReachPoint
+from wfa_planning_evaluation_framework.simulator.privacy_tracker import (
+    DP_NOISE_MECHANISM_LAPLACE,
+)
+from wfa_planning_evaluation_framework.simulator.privacy_tracker import NoisingEvent
+from wfa_planning_evaluation_framework.simulator.privacy_tracker import PrivacyTracker
+from wfa_planning_evaluation_framework.simulator.publisher import Publisher
+from wfa_planning_evaluation_framework.simulator.simulation_parameters import (
+    SimulationParameters,
+)
+
+
+class HaloSimulator:
+    def __init__(
+        self,
+        data_set: DataSet,
+        params: SimulationParameters,
+        privacy_tracker: PrivacyTracker,
+    ):
+        """Halo simulator.
+
+        Args:
+          data_set:  The data that will be used as ground truth for this
+            halo instance.
+          params:  Configuration parameters for this halo instance.
+          privacy_tracker: A PrivacyTracker object that will be used to track
+            privacy budget consumption for this simulation.
+        """
+        self._data_set = data_set
+        self._params = params
+        self._privacy_tracker = privacy_tracker
+        self._publishers = []
+        for i in range(data_set.publisher_count):
+            self._publishers.append(
+                Publisher(data_set._data[i], i, params, privacy_tracker)
+            )
+
+    def true_reach_by_spend(
+        self, spends: List[float], max_frequency: int = 1
+    ) -> ReachPoint:
+        """Returns the true reach obtained for a given spend vector.
+
+        Args:
+            spends:  The hypothetical amount spend vector, equal in length to
+              the number of publishers.  spends[i] is the amount that is
+              spent with publisher i.
+            max_frequency:  The maximum frequency for which to report reach.
```

Comment:
I'm not sure how to read this. Can you explain further?

Reply:
Added a comment.